include/matrix_expression_iterator.h

```diff
@@ -0,0 +1,156 @@
+#pragma once
+
+#include <iterator>
+#include "access_category.h"
+
+namespace AMatrix {
+
+template <typename TDataType, typename TExpressionType, std::size_t TCategory>
+class ExpressionRandomAccessIterator
+    : public std::iterator<std::random_access_iterator_tag, TDataType> {
```

Comment:
I have recently saw that the std::iterator will be deprecated in c++17. The idea is to put the using needs by iterator_trait directly in the class:

```c++
       public:
        using iterator_category = std::forward_iterator_tag;
        using value_type = TDataType;
        using difference_type = std::size_t;
        using pointer = TDataType*;
        using reference = TDataType&;
```